### PR TITLE
Support getting the available aggregation options

### LINKED ISSUE #28702: [MLv2] Implement functions for available aggregations
To power this UI:

image

Needs to return available aggregation types for the current database as a list of maps in JS and UI-friendly (i.e., `clj->js` manner) manner, and we should be able to take one of these maps and pass it back in to `(aggregation query ...)` (as added by #28693) to add it to the query 

For aggregations that need a Field or expression, either include the possible Fields inline, or implement a separate function to get the options for a choice from the first step.

We'll have to make sure we include Fields that aren't currently projected e.g. if we already have a `[:sum X]` aggregation it should still be possible to add a `[:sum Y]`, even if `Y` isn't currently in the `:lib/stage-metadata` for the current stage

### PR DESCRIPTION
**Pull request by metamben**
Fixes #28702.

The supported usage is demonstrated in `metabase.lib.aggregation-test/aggregation-operator-test`:
1. `available-aggregation-operators` is called on the query (and optionally the stage number),
1. `display-info` is used on the elements of the returned list to present the operators,
1. `aggregation-operator-columns` is called on the selected operator (if it requires columns),
1. `display-info` is called on the the possible columns to present them,
1. `aggregation-clause` is called on the selected operator (and column, if one is selected),
1. `aggregate` is called to add the returned aggregation clause ito the query.

The PR also resolves a number of minor issues found in the review of another PR (sorry).

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/30836)

### CODE REVIEW

File: shared/src/metabase/shared/util/time.cljc [hunk 1]
```diff
@@ -9,12 +9,15 @@
...
-(shared.ns/import-fn internal/day-diff)
+(shared.ns/import-fns
+ [common
+  to-range]
+ [internal
+  valid?
+  same-day?
+  same-month?
+  same-year?
+  day-diff])
```
Comment (metamben) [inline shared/src/metabase/shared/util/time.cljc:20]: This is an unrelated improvement.

File: src/metabase/lib/field.cljc [hunk 2]
```diff
@@ -287,7 +287,7 @@
   [query stage-number field-ref]
   (lib.temporal-bucket/available-temporal-buckets query stage-number (resolve-field-metadata query stage-number field-ref)))
 
-(defn- fingerprint-based-default [fingerprint]
+(defn- fingerprint-based-default-unit [fingerprint]
```
Comment (metamben) [inline src/metabase/lib/field.cljc:290]: Another unrelated improvement.

File: src/metabase/lib/stage.cljc [hunk 3]
```diff
@@ -154,10 +154,14 @@
...
-            :lib/source-column-alias  (:name expression)
-            :lib/desired-column-alias (unique-name-fn (:name expression))))))
+     (let [base-type (:base-type expression)]
+       (cond-> (assoc expression
+                      :lib/source               :source/expressions
+                      :lib/source-column-alias  (:name expression)
+                      :lib/desired-column-alias (unique-name-fn (:name expression)))
+         (and (not (:effective-type expression))
+              base-type)
+         (assoc :effective-type base-type))))))
```
Comment (metamben) [inline src/metabase/lib/stage.cljc:164]: My understanding is that we don't want the FE to rely on `base-type`, `effective-type` should always be defined.

File: test/metabase/lib/aggregation_test.cljc [hunk 4]
```diff
@@ -244,6 +244,170 @@
...
+                   :lib/source :source/expressions}
+                  {:display-name "ID",
+                   :effective-type :type/BigInteger,
+                   :semantic-type :type/PK,
+                   :lib/source :source/implicitly-joinable}
+                  {:display-name "Name",
+                   :effective-type :type/Text,
+                   :semantic-type :type/Name,
+                   :lib/source :source/implicitly-joinable}]
+        scope-cols all-cols
```
Comment (metamben) [inline test/metabase/lib/aggregation_test.cljc:305]: For some unknown to me reason, the `scope` type, which is required for `min` and `max`, includes the `category` type which in turn includes booleans. This is unexpected, because we say that booleans are not ordered.

The current behavior (stats.metabase.com) is that min and max aggregations are offered for booleans but running the query fails with `function max(boolean) does not exist`.

File: test/metabase/lib/metadata/jvm_test.clj [hunk 5]
```diff
@@ -83,34 +83,3 @@
...
-              {:unit :day-of-week}
-              {:unit :day-of-month}
-              {:unit :day-of-year}
-              {:unit :week-of-year}
-              {:unit :month-of-year}
-              {:unit :quarter-of-year}]
-             (->> (lib.metadata.calculation/metadata query)
-                  first
-                  (lib/available-temporal-buckets query)
-                  (mapv #(select-keys % [:unit :default]))))))))
```
Comment (metamben) [inline test/metabase/lib/metadata/jvm_test.clj:116]: Unrelated improvement: the test was made CLJC compatible and moved to `metabase.lib.temporal-bucket-test`.

File: src/metabase/lib/aggregation.cljc [hunk 6]
```diff
@@ -226,3 +229,75 @@
...
+
+(defmethod lib.metadata.calculation/display-name-method :mbql.aggregation/operator
+  [_query _stage-number {:keys [display-info]} _display-name-style]
+  (:display-name (display-info)))
+
+(defmethod lib.metadata.calculation/display-info-method :mbql.aggregation/operator
+  [_query _stage-number {:keys [display-info requires-field?] short-name :short}]
+  (assoc (display-info)
+         :short short-name
+         :requires-field requires-field?))
```
Comment (kulyk) [inline src/metabase/lib/aggregation.cljc:247]: Would it make sense to use `requires-column` instead of `requires-field` according to #30060 ?

**Comment by kulyk** [comment]
@metamben I noticed that when I add a "count" aggregation, the final query looks a bit different from MLv1:

```js
// MLv1
{
  aggregation: [ ["count"] ]
}

// MLv2 (this branch), notice the `null`
{
  aggregation: [ ["count", null] ]
}
```

I haven't tried it yet, but I'm worried FE MLv1 could treat it as invalid. Not yet sure if it should be fixed, but wanted to bring this to your attention just in case

**Review by kulyk — APPROVED** [review]
Works great! 👏

**Review by snoe — APPROVED** [review]
LGTM
